src/history_manager.py

```diff
@@ -95,71 +129,53 @@ def rewind_history(self):
 	# Serialized operations ####################################################
 
 	def add_operation(self, operation):
+
 		self._image.set_surface_as_stable_pixbuf()
-		# print('add_operation_to_history')
-		# print(operation['tool_id'])
-		# if 'select' in operation['tool_id']:
-		# 	print(operation['operation_type'])
-		# 	print('-----------------------------------')
+		if self._undo_history[-1].is_full():
+			self.add_state(self._image.main_pixbuf.copy())
+
+		self._undo_history[-1].add_operation(operation)
 		self._is_saved = False
-		self._undo_history.append(operation)
+		self._image.update_title()
+		
 
 	############################################################################
 	# Cached pixbufs ###########################################################
 
 	def set_initial_operation(self, rgba_array, pixbuf, width, height):
+		
 		r = float(rgba_array[0])
 		g = float(rgba_array[1])
 		b = float(rgba_array[2])
 		a = float(rgba_array[3])
-		self.initial_operation = {
+		initial_operation = {
 			'tool_id': None,
 			'pixbuf': pixbuf,
 			'rgba': Gdk.RGBA(red=r, green=g, blue=b, alpha=a),
 			'width': width, 'height': height
 		}
+		self._undo_history.append(State(initial_operation))
 
 	def add_state(self, pixbuf):
 		if pixbuf is None:
 			# Context: an error message
-			raise Exception(_("Attempt to save an invalid state"))
-		self._undo_history.append({
+			raise Exception("Attempt to save an invalid state")
+		
+		new_state = {
 			'tool_id': None,
 			'pixbuf': pixbuf,
 			'width': pixbuf.get_width(),
 			'height': pixbuf.get_height()
-		})
+		}
+
+		self._undo_history.append(State(new_state))
 		self._is_saved = True
 
 	def has_initial_pixbuf(self):
-		return self.initial_operation['pixbuf'] is not None
-
-	def get_last_saved_state(self):
-		index = self._get_last_state_index(False)
-		if index == -1:
-			return self.initial_operation
-		else:
-			return self._undo_history[index]
+		return len(self._undo_history) > 0
 
-	def _get_last_state_index(self, allow_yeeting_states):
-		"""Return the index of the last "state" operation (dict whose 'tool_id'
-		value is None) in the undo-history. If there is no such operation, the
-		returned index is -1 which means the only known state is the
-		self.initial_operation attribute."""
-
-		returned_index = -1
-		nbPixbufs = 0
-		for op in self._undo_history:
-			if op['tool_id'] is None:
-				last_saved_pixbuf_op = op
-				returned_index = self._undo_history.index(op)
-				nbPixbufs += 1
-
-		# TODO if there are too many pixbufs in the history, remove a few ones
-		# Issue #200, needs more design because saving can change the data
```

Comment:
please restore this one too, it's a comment related to the design of the history but not regarding performance, it's a distinct bug

Reply:
__get_last_state_index_ is no longer there, so I moved this comment (and the other one) to the _add_operation_ method.